### PR TITLE
Split up NLSolve.jl and introduce IsFiniteExcpetion

### PR DESCRIPTION
**Pull request by KristofferC**
First commit fixes the bloat in the file `NLsolve.jl`. There were many type definitions and unrelated function definitions there so I separated them out in a few files

Second commit introduces an "IsFiniteException" that is thrown when a non finite value is encountered in the function value. Before, this was performedn in a `error("...")` block but using proper exceptions is more idiomatic. I also included tests for the exceptions.

### CODE REVIEW

File: src/newton.jl [hunk 1]
```diff
@@ -96,8 +92,6 @@ function newton_{T}(df::AbstractDifferentiableMultivariateFunction,
         df.g!(x, fjac)
         g_calls += 1
 
-        g = fjac'*fvec
```
Comment (KristofferC) [inline src/newton.jl]: Forgot that I added this. g is not used except in the catch block so move it in there to save a matvec multiplication

**Comment by spencerlyon2** [comment]
Hey  @KristofferC with the other PRs that have been merged, what is the status of this PR?

**Comment by KristofferC** [comment]
I just need to rebase it.

**Comment by KristofferC** [comment]
Rebased.

**Comment by spencerlyon2** [comment]
Thanks! I think we are ready for a METADATA re-tag now

**Comment by KristofferC** [comment]
Yes, after #34 is merged I have no immediate plans on adding or changing anything so a re-tag would be great.

**Pull request merged**